### ISSUE TITLE
Discount amount changing upon updating order

### ISSUE THREAD

**Issue by matty0501** [woocommerce/woocommerce#24633]
**Describe the bug**
I have discovered an issue where, under certain conditions, the discount amount on an order is changing when the order is updated. For example, I have a product that is £8.99 (including 20% tax) shipping which is £9.59 (including 20% tax), and a coupon which gives £5 discount. After checkout, the order is showing a discount of £5 (£4.17 excl tax). If I go into the order on the admin screen, and click update, the order will change to show a discount of £4.99 (£4.16 excl tax). I have not made any changes to the order, simply just clicked on 'Update'. No other values change, so now the order values do not add up to the order total.

I understand that it is only out by a penny, but customers are getting confused as they applied a £5 coupon and received an email that shows this on the order, but when the order is updated it shows as a £4.99 discount.

In terms of tax settings, I am inputting prices inclusive of tax and displaying prices inclusive of tax.

I discovered this issue on my production site, and have replicated the issue on a new staging site that I spun up. The staging site is using the default Twenty Nineteen theme and only has WooCommerce installed.

**Screenshots**
Order on customer My Account before it is updated:
![Account Before Update](https://user-images.githubusercontent.com/17788098/64908425-18557d00-d6f8-11e9-999f-414e30501a67.png)

Order on Admin before it is updated:
![Admin Before Update](https://user-images.githubusercontent.com/17788098/64908428-2905f300-d6f8-11e9-9adb-9036d741df73.png)

Order on customer My Account after it is updated:
![Account After Update](https://user-images.githubusercontent.com/17788098/64908444-5a7ebe80-d6f8-11e9-9f64-47018adb9cf4.png)

Order on Admin after it is updated:
![Admin After Update](https://user-images.githubusercontent.com/17788098/64908446-65395380-d6f8-11e9-9ffd-ab819d0f8db6.png)

**Expected behavior**
I would expect the discount amount not to change if I have not edited the discount on the order. If a customer uses a £5 coupon, the discount should show as £5.

**Isolating the problem (mark completed items with an [x]):**
- [X ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X ] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
`
### WordPress Environment ###

WordPress address (URL): http://157.245.37.33
Site address (URL): http://157.245.37.33
WC Version: 3.7.0
REST API Version: ✔ 1.0.2
Log Directory Writable: ✔
WP Version: 5.2.3
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_GB
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.29 (Ubuntu)
PHP Version: 7.2.19-0ubuntu0.18.04.2
PHP Post Max Size: 16 MB
PHP Time Limit: 60
PHP Max Input Vars: 1000
cURL Version: 7.58.0
OpenSSL/1.1.1

SUHOSIN Installed: –
MySQL Version: 5.7.27-0ubuntu0.18.04.1
Max Upload Size: 16 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.7.0
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 3.03MB
Database Data Size: 2.06MB
Database Index Size: 0.97MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 1.48MB + Index: 0.02MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 3.7.0

### Inactive Plugins (2) ###

Akismet Anti-Spam: by Automattic – 4.1.2
Hello Dolly: by Matt Mullenweg – 1.7.2

### Settings ###

API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #6 - /
Basket: #7 - /basket/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Nineteen
Version: 1.4
Author URL: https://en-gb.wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build
personally we recommend using a child theme. See: How to create a child theme

WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 1
Oldest: 2019-09-14 12:52:12 +0100
Newest: 2019-09-14 12:52:12 +0100

Pending: 0
Oldest: –
Newest: –

Canceled: 0
Oldest: –
Newest: –

In-progress: 0
Oldest: –
Newest: –

Failed: 0
Oldest: –
Newest: –

`
</details>

**Comment by matty0501** [comment]
I've had a look through the code, and I'm not sure if this behaviour is caused by how WooCommerce calculates the discount total on line 1513 of /woocommerce/includes/abstracts/abstract-wc-order.php

`$this->set_discount_total( $cart_subtotal - $cart_total );`

Looking at lines 1468 to 1471, these values are defined as:
```
foreach ( $this->get_items() as $item ) {
			$cart_subtotal += round( $item->get_subtotal(), wc_get_price_decimals() );
			$cart_total    += round( $item->get_total(), wc_get_price_decimals() );
}
```

As this is calculating the discount total using the total and subtotal of each product based on the number of decimal places shown in the store (2 in my case), it's producing rounding errors.

Using the figures from the example in my post above with a tax rate of 20%, the product is £8.99 inc tax, so £7.49166 exc tax. After the £5 discount is applied, the product is £3.99 inc tax, so £3.325 exc tax.

Because WC is calculating to 2 decimal places (as set in the price display setting), it's calculating 7.49-3.33 = 4.16, which is wrong as this will show as a discount of £4.99 including tax, not the £5 that was applied.

However, if we calculate to three decimal places, the product is £7.492 excl tax before the discount, and should be £3.325 after the discount. So 7.492-3.325 = 4.167, which would round to 4.17, or £5 including tax.

I'm not the best at PHP, so the above might not be where the issue lies, but I wanted to share what I found in case it helps. Apologies for all of the maths, I hope that I explained it in a way that was easy enough to follow.

**Comment by matty0501** [comment]
In support of my above post regarding what I think is causing this issue, I tried hooking into the calculate_totals function where the code above lives:

At the start of the function is the hook:
`do_action( 'woocommerce_order_before_calculate_totals', $and_taxes, $this );`
And at the end of the function is the hook:
`do_action( 'woocommerce_order_after_calculate_totals', $and_taxes, $this );`

What I did was hook into the start of the function and essentially change wc_get_price_decimals() to return 4 (so rounding is done to 4 decimal places). I then hooked into the end of the function to remove this so that prices continue to be displayed to 2 decimal places on the site.

And it worked, now when updating the order it preserves the correct discount amount of £5.00, or £4.17 excluding tax!

Full code that I used is:
```
add_action( 'woocommerce_order_before_calculate_totals', 'matt_change_decimals' );
function matt_change_decimals(){	
	add_filter( 'wc_get_price_decimals', 'matt_return_decimals' );	
}

add_action( 'woocommerce_order_after_calculate_totals', 'matt_remove_change_decimals' );
function matt_remove_change_decimals(){	
	remove_filter( 'wc_get_price_decimals', 'matt_return_decimals' );	
}

function matt_return_decimals() {	
	return 4;	
}
```
